### ISSUE TITLE
Nil relations when parsing joined results

### ISSUE THREAD

**Issue by zauner** [imdrasil/jennifer.cr#129]
Hi!

It seems there is an issue when converting result sets from joined queries to Jennifer.cr objects. If relations from records from a result point to _the same_ referenced record, it is correctly assigned only once. All subsequent occurences are nil.

Here's a test case which illustrates the problem:

Data (Postgres):
```SQL
-- Users
CREATE TABLE test_users (
    id bigint NOT NULL,
    name character varying(50)
);
INSERT INTO test_users (id, name) VALUES (1, 'Anna');
INSERT INTO test_users (id, name) VALUES (2, 'Bert');
INSERT INTO test_users (id, name) VALUES (3, 'Charly');

-- Posts
CREATE TABLE test_posts (                                                                  
    id bigint NOT NULL,                                                                    
    user_id bigint NOT NULL,                                                               
    content character varying NOT NULL                                                     
);                                                                                                                                                                            
INSERT INTO test_posts (id, user_id, content) VALUES (1, 1, 'First Post');                 
INSERT INTO test_posts (id, user_id, content) VALUES (2, 1, 'Second Post');                
INSERT INTO test_posts (id, user_id, content) VALUES (3, 2, 'Third Post');                 
INSERT INTO test_posts (id, user_id, content) VALUES (4, 1, 'Fourth Post');                
```

Models:
```Crystal
class TestUser < Jennifer::Model::Base

  mapping(
    id: Primary64,
    name: String
  )

  has_many :posts, TestPost, foreign: "user_id"

end

class TestPost < Jennifer::Model::Base

  mapping(
    id: Primary64,
    user_id: Int64,
    content: String
  )

  belongs_to :user, TestUser, foreign: "user_id"

end
```

First, I load _all users with their posts_ -- everything looks good:
```Crystal
TestUser.all.relation("posts").with("posts").each do |u|
  puts "User: #{u.name}"
  puts "Posts:"
  u.posts.each do |p|
    puts "* #{p.content}"
  end
  puts "\n"
end

# 2018-04-12 19:44:43 +00:00: 854 µs SELECT test_users.*, test_posts.* 
# FROM test_users                                                      
# LEFT JOIN test_posts ON test_posts.user_id = test_users.id           
                                                                     
# User: Anna                                                           
# Posts:                                                               
# * First Post                                                         
# * Second Post                                                        
# * Fourth Post                                                        
                                                                     
# User: Bert                                                           
# Posts:                                                               
# * Third Post                                                         
                                                                     
# User: Charly                                                         
# Posts:                                                               
```

However, when loading _all posts with their users_ `TestPost#user` gets `nil`, when the user was already referenced by a previous post:
```Crystal
TestPost.all.relation("user").with("user").each do |p|
  puts "* #{p.content}"
  if p.user
    puts "by #{p.user!.name}"
  else
    puts "BOOOOOOM! Caught nil."
  end
  puts "\n"
end

# 2018-04-12 19:44:43 +00:00: 487 µs SELECT test_posts.*, test_users.*
# FROM test_posts                                                     
# LEFT JOIN test_users ON test_users.id = test_posts.user_id          
                                                                    
# * First Post                                                        
# by Anna                                                             
                                                                    
# * Second Post                                                       
# BOOOOOOM! Caught nil.                                               
                                                                    
# * Third Post                                                        
# by Bert                                                             
                                                                    
# * Fourth Post                                                       
# BOOOOOOM! Caught nil.                         
```

When I ommit the relation joins and go with just `TestPost.all.each`, it works as expected.

I hope that's enough to explain the problem, let me know if I can add any more information.

**Comment by imdrasil** [comment]
I guess I know where is the root cause. Also, prefer using `#eager_load` and `#includes` over `#relation` + `#with`.

**Comment by zauner** [comment]
Cool, can confirm that this is solved for me in current v0.6.0 branch, thanks!
